nullaway/src/test/java/com/uber/nullaway/EnsuresNonNullIfTests.java

```diff
@@ -0,0 +1,898 @@
+    defaultCompilationHelper
+        .addSourceLines(
+            "Foo.java",
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "import com.uber.nullaway.annotations.EnsuresNonNullIf;",
+            "class Foo {",
+            "  @Nullable Item nullableItem;",
+            "  @EnsuresNonNullIf(value=\"nullableItem\", result=true)",
+            "  public boolean hasNullableItem() {",
+            "    // BUG: Diagnostic contains: Method is annotated with @EnsuresNonNullIf but does not ensure fields [nullableItem]",
+            "    return true || nullableItem != null;",
+            "  }",
+            "  public int runOk() {",
+            "    if(!hasNullableItem()) {",
+            "      return 1;",
+            "    }",
+            "    nullableItem.call();",
+            "    return 0;",
+            "  }",
+            "}")
+        .addSourceLines(
+            "Item.java", "package com.uber;", "class Item {", "  public void call() { }", "}")
+        .doTest();
+  }
+
+  @Test
+  public void noSemanticIssue_combinationOfExpressionAndLiteralBoolean() {
+    defaultCompilationHelper
+        .addSourceLines(
+            "Foo.java",
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "import com.uber.nullaway.annotations.EnsuresNonNullIf;",
+            "class Foo {",
+            "  @Nullable Item nullableItem;",
+            "  @EnsuresNonNullIf(value=\"nullableItem\", result=true)",
+            "  public boolean hasNullableItem() {",
+            "    return true && nullableItem != null;",
+            "  }",
+            "  public int runOk() {",
+            "    if(!hasNullableItem()) {",
+            "      return 1;",
+            "    }",
+            "    nullableItem.call();",
+            "    return 0;",
+            "  }",
+            "}")
+        .addSourceLines(
+            "Item.java", "package com.uber;", "class Item {", "  public void call() { }", "}")
+        .doTest();
+  }
+
+  @Test
+  public void semanticIssues_methodDeclarationReversed() {
+    defaultCompilationHelper
+        .addSourceLines(
+            "Foo.java",
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "import com.uber.nullaway.annotations.EnsuresNonNullIf;",
+            "class Foo {",
+            "  @Nullable Item nullableItem;",
+            "  public int runOk() {",
+            "    if(!hasNullableItem()) {",
+            "      return 1;",
+            "    }",
+            "    nullableItem.call();",
+            "    return 0;",
+            "  }",
+            "  @EnsuresNonNullIf(value=\"nullableItem\", result=true)",
+            "  public boolean hasNullableItem() {",
+            "    if(nullableItem != null) {",
+            "      // BUG: Diagnostic contains: The method ensures the non-nullability of the fields, but returns incorrectly",
```

Comment:
I don't think this should be an error.  The annotation means, "If this method returns `true`, then `nullableItem` is guaranteed to be non-null."  If the method returns `false`, then the annotation says nothing about whether `nullableItem` will be null or not.  And at callers, we should only be using nullability information from the annotation for the case where the returned boolean matches the `result` field of the annotation.  So there is no need to perform any check within the method when we know it is not returning the `result` value.  The error reported for the `return true` statement is correct.  Do you agree?

Reply:
That does make sense to me! Let me work on it!

Reply:
Here: https://github.com/uber/NullAway/pull/1044/commits/56a0984a935cc89a0dd7819590afe6c7f4bd481e